openwisp_controller/config/whois/utils.py

```diff
@@ -34,3 +73,89 @@ def setUp(self):
         OrganizationConfigSettings.objects.create(
             organization=self._get_org(), whois_enabled=True
         )
+
+
+class WHOISTransactionMixin:
+    @staticmethod
+    def _mocked_client_response():
+        mock_response = mock.MagicMock()
+        mock_response.city.name = "Mountain View"
+        mock_response.country.name = "United States"
+        mock_response.continent.name = "North America"
+        mock_response.postal.code = "94043"
+        mock_response.traits.autonomous_system_organization = "Google LLC"
+        mock_response.traits.autonomous_system_number = 15169
+        mock_response.traits.network = "172.217.22.0/24"
+        mock_response.location.time_zone = "America/Los_Angeles"
+        mock_response.location.latitude = 2
+        mock_response.location.longitude = 23
+        return mock_response
+
+    def _task_called(self, mocked_task, task_name="WHOIS lookup"):
+        org = self._get_org()
+
+        with self.subTest(f"{task_name} task called when last_ip is public"):
+            with mock.patch("django.core.cache.cache.set") as mocked_set:
+                device = self._create_device(last_ip="172.217.22.14")
+                mocked_task.assert_called()
+                mocked_set.assert_called_once()
+        mocked_task.reset_mock()
+
+        with self.subTest(
+            f"{task_name} task called when last_ip is changed and is public"
+        ):
+            with mock.patch("django.core.cache.cache.get") as mocked_get:
+                device.last_ip = "172.217.22.10"
+                device.save()
+                mocked_task.assert_called()
+                # The cache `get` is called twice, once for `whois_enabled` and
+                # once for `approximate_location_enabled`
+                mocked_get.assert_called()
+        mocked_task.reset_mock()
+
+        with self.subTest(f"{task_name} task not called when last_ip is private"):
+            device.last_ip = "10.0.0.1"
+            device.save()
+            mocked_task.assert_not_called()
+        mocked_task.reset_mock()
+
+        with self.subTest(f"{task_name} task not called when WHOIS is disabled"):
+            Device.objects.all().delete()
+            org.config_settings.whois_enabled = False
+            # Invalidates old org config settings cache
+            org.config_settings.save(update_fields=["whois_enabled"])
+            device = self._create_device(last_ip="172.217.22.14")
+            mocked_task.assert_not_called()
+        mocked_task.reset_mock()
+
+        with self.subTest(
```

Comment:
Add one more check if it is not there, task not called via DeviceChecksumView when WHOIS is disabled